File: app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/model/Weekdays.java
package de.westnordost.streetcomplete.quests.opening_hours.model;

import android.content.res.Resources;

import java.text.DateFormatSymbols;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import de.westnordost.streetcomplete.R;

public class Weekdays
{
	// in ISO 8601 order
	private static final String[] OSM_ABBR_WEEKDAYS = {"Mo","Tu","We","Th","Fr","Sa","Su","PH"};
	private static final int PUBLIC_HOLIDAY = 7;
	private static final int WEEKDAYS_VALUES = OSM_ABBR_WEEKDAYS.length;

	private static final NumberSystem WEEKDAY_NUMBER_SYSTEM = new NumberSystem(0, 6);

	private boolean[] data = new boolean[WEEKDAYS_VALUES];

	public static String[] getNames(Resources r)
	{
		DateFormatSymbols symbols = DateFormatSymbols.getInstance();
		String[] result = Arrays.copyOf(toIso8601Order(symbols.getWeekdays()),WEEKDAYS_VALUES);
		result[PUBLIC_HOLIDAY] = r.getString(R.string.quest_openingHours_public_holidays);
		return result;
	}

	public static String[] getShortNames(Resources r)
	{
		DateFormatSymbols symbols = DateFormatSymbols.getInstance();
		String[] result = Arrays.copyOf(toIso8601Order(symbols.getShortWeekdays()),WEEKDAYS_VALUES);
		result[PUBLIC_HOLIDAY] = r.getString(R.string.quest_openingHours_public_holidays_short);
		return result;
	}

	public static int getWeekdayIndex(String name)
	{
		return Arrays.asList(OSM_ABBR_WEEKDAYS).indexOf(name);
	}

	private static String[] toIso8601Order(String[] javaCalendarOrder)
	{
		String[] result = new String[7];
		int shift = 1;
		for(int i = 0; i < 7; ++i)
		{
			result[i] = javaCalendarOrder[1+(i+shift)%7];
		}
		return result;
	}

	public Weekdays()
	{

	}

	public Weekdays(boolean[] selection)
	{
		for(int i=0; i<selection.length && i <data.length; ++i)
		{
			data[i] = selection[i];
		}
	}

	public boolean[] getSelection()
	{
		return Arrays.copyOf(data, data.length);
	}

	@Override public String toString()
	{
		return toStringUsing(OSM_ABBR_WEEKDAYS, ",", "-");
	}

	public String toLocalizedString(Resources r)
	{
		return toStringUsing(Weekdays.getShortNames(r), ", ", "–");
	}

	private String toStringUsing(String[] names, String seperator, String range)
	{
		StringBuilder sb = new StringBuilder();
		boolean first = true;

		for(CircularSection section : getWeekdaysAsCircularSections())
		{
			if(!first)	sb.append(seperator);
			else		first = false;

			sb.append(names[section.getStart()]);
			if (section.getStart() != section.getEnd())
			{
				// i.e. Mo-We
				if(WEEKDAY_NUMBER_SYSTEM.getSize(section) > 2)
				{
					sb.append(range);
				}
				// Mo,Tu
				else
				{
					sb.append(seperator);
				}
				sb.append(names[section.getEnd()]);
			}
		}

		// the rest (special days). Currently only "PH"
		for(int i = 7; i < data.length; ++i)
		{
			if(!data[i]) continue;

			if(!first)	sb.append(seperator);
			else		first = false;

			sb.append(names[i]);
		}

		return sb.toString();
	}

	private List<CircularSection> getWeekdaysAsCircularSections()
	{
		List<CircularSection> result = new ArrayList<>();
		Integer currentStart = null;
		for(int i = 0; i < 7; ++i)
		{
			if(currentStart == null)
			{
				if(data[i]) currentStart = i;
			}
			else
			{
				if(!data[i])
				{
					result.add(new CircularSection(currentStart, i-1));
					currentStart = null;
				}
			}
		}
		// section that goes until the end
		if(currentStart != null)
		{
			result.add(new CircularSection(currentStart, 6));
		}

		return WEEKDAY_NUMBER_SYSTEM.merged(result);
	}

	public boolean intersects(Weekdays other)
	{
		for (int i = 0; i < data.length; i++)
		{
			if(data[i] && other.data[i]) return true;
		}
		return false;
	}

	public boolean isSelectionEmpty()
	{
		for (boolean day : data)
		{
			if (day) return false;
		}
		return true;
	}

	@Override public boolean equals(Object o)
	{
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		return Arrays.equals(data, ((Weekdays) o).data);
	}

	@Override public int hashCode()
	{
		return Arrays.hashCode(data);
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours/model/Weekdays.kt
package de.westnordost.streetcomplete.quests.opening_hours.model

import android.content.res.Resources

import java.text.DateFormatSymbols

import de.westnordost.streetcomplete.R

class Weekdays {

    private val data = BooleanArray(OSM_ABBR_WEEKDAYS.size)

    val selection: BooleanArray get() = data.copyOf()

    constructor()

    constructor(selection: BooleanArray) {
        var i = 0
        while (i < selection.size && i < data.size) {
            data[i] = selection[i]
            ++i
        }
    }

    fun isSelectionEmpty() = data.all { !it }

    override fun toString() = toStringUsing(OSM_ABBR_WEEKDAYS, ",", "-")

    fun toLocalizedString(r: Resources) = toStringUsing(getShortNames(r), ", ", "–")

    fun toStringUsing(names: Array<String>, separator: String, range: String): String {
        val sb = StringBuilder()
        var first = true

        for (section in toCircularSections()) {
            if (!first)
                sb.append(separator)
            else
                first = false

            sb.append(names[section.start])
            if (section.start != section.end) {
                // i.e. Mo-We vs Mo,Tu
                sb.append(if (WEEKDAY_NUMBER_SYSTEM.getSize(section) > 2) range else separator)
                sb.append(names[section.end])
            }
        }

        // the rest (special days). Currently only "PH"
        for (i in 7 until data.size) {
            if (!data[i]) continue

            if (!first)
                sb.append(separator)
            else
                first = false

            sb.append(names[i])
        }

        return sb.toString()
    }

    // section that goes until the end
    private fun toCircularSections(): List<CircularSection> {
        val result = mutableListOf<CircularSection>()
        var currentStart: Int? = null
        for (i in 0..6) {
            if (currentStart == null) {
                if (data[i]) currentStart = i
            } else {
                if (!data[i]) {
                    result.add(CircularSection(currentStart, i - 1))
                    currentStart = null
                }
            }
        }
        if (currentStart != null) {
            result.add(CircularSection(currentStart, 6))
        }

        return WEEKDAY_NUMBER_SYSTEM.merged(result)
    }

    fun intersects(other: Weekdays): Boolean {
        for (i in data.indices) {
            if (data[i] && other.data[i]) return true
        }
        return false
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Weekdays) return false
        return data.contentEquals(other.data)
    }

    override fun hashCode() = data.contentHashCode()

    companion object {
        // in ISO 8601 order
        val OSM_ABBR_WEEKDAYS = arrayOf("Mo", "Tu", "We", "Th", "Fr", "Sa", "Su", "PH")
        private const val PUBLIC_HOLIDAY = 7

        private val WEEKDAY_NUMBER_SYSTEM = NumberSystem(0, 6)

        fun getNames(r: Resources): Array<String> {
            val symbols = DateFormatSymbols.getInstance()
            val result = symbols.weekdays.toIso8601Order().copyOf(OSM_ABBR_WEEKDAYS.size)
            result[PUBLIC_HOLIDAY] = r.getString(R.string.quest_openingHours_public_holidays)
            return result.requireNoNulls()
        }

        fun getShortNames(r: Resources): Array<String> {
            val symbols = DateFormatSymbols.getInstance()
            val result = symbols.shortWeekdays.toIso8601Order().copyOf(OSM_ABBR_WEEKDAYS.size)
            result[PUBLIC_HOLIDAY] = r.getString(R.string.quest_openingHours_public_holidays_short)
            return result.requireNoNulls()
        }

        fun getWeekdayIndex(name: String) = OSM_ABBR_WEEKDAYS.indexOf(name)

        private fun Array<String>.toIso8601Order() = Array(7) { this[1 + (it + 1) % 7] }
    }
}
